### PR TITLE
Enable oneDNN implementation in LSTM op

### PR DESCRIPTION
**Pull request by yanbing-j**
### Description
This PR is to enable oneDNN implementation in LSTM op to improve the performance of it. Both FP32 and BF16 are supported.

### Performance improvement
In CPX 28C, with setting iomp and jemalloc.
We choose 8 LSTM input options (including input_size, hidden_size, num_layers, bidirectional, bias, batch_first, dropout, batch_size, seq_len), and the final option is a real input from train-clean-100 in LibriSpeech dataset. The performance improvements are shown in the following figures. We can see that LSTM with oneDNN implementation can perform better than the original.

In single socket:
image

image

In single core:
image

image

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -2310,6 +2310,46 @@ def _cudnn_rnn(
...
+    output = input.new_empty(out_shape)
+    if hx_ is None:
+        hy = torch.empty(0, device=input.device)
+    else:
+        hy = hx_.new_empty(hx_.shape)
+    if cx_ is None:
+        cy = torch.empty(0, device=input.device)
+    else:
+        cy = cx_.new_empty(cx_.shape)
+    workspace = input.new_empty([hidden_size * 1024], dtype=torch.uint8)
```
Comment (jgong5) [inline torch/_meta_registrations.py:2349]: I guess `workspace` doesn't matter here, just creating an empty tensor would be good enough?

File: aten/src/ATen/native/mkldnn/RNN.cpp [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+// Init a aten tensor according to ideep tensor's desc.
+static inline Tensor empty_aten_tensor_from_desc(
+    const ideep::tensor::desc& desc,
+    const at::TensorOptions& options) {
+  auto ndims = desc.data.ndims;
+  auto nblks = desc.blocking_desc().inner_nblks;
+  std::vector<int64_t> at_sizes(ndims + nblks);
+  auto padded_dims = desc.padded_dims();
+  auto blk_sizes = desc.blocking_desc().inner_blks;
+  auto blk_idxs = desc.blocking_desc().inner_idxs;
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/RNN.cpp:245]: Do we really have to parse the internal blocking descriptors of onednn to get the workspace aten tensor? Can we just model it as a 1D tensor buffer from aten side?

File: aten/src/ATen/native/mkldnn/RNN.cpp [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+    int64_t num_layers, bool has_biases, bool batch_first, double dropout_p,
+    bool train, bool bidirectional, IntArrayRef batch_sizes) {
+//   TORCH_CHECK(!train || dropout_p == 0.0, "mkldnn_rnn doesn't support dropout");
+  TORCH_CHECK(batch_sizes.size() == 0, "mkldnn_rnn doesn't support packed input");
+  if (static_cast<ideep::rnn_kind>(mode) != ideep::rnn_kind::LSTM) {
+    TORCH_CHECK(!cx_.defined(), "mkldnn_rnn: illegal defined cx for non-LSTM RNN");
+  }
+
+  auto input = input_;
+  bool is_input_packed = batch_sizes.size() != 0;
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/RNN.cpp:508]: This is always false, why bother check?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: test/test_mkldnn.py [hunk 4]
```diff
@@ -41,6 +43,9 @@ def has_bf16_support():
 
 types = [torch.float, torch.bfloat16]
 
+def get_rand_seed():
+    return int(time.time() * 1000000000)
```
[the remark this answers is not in the document]
Comment (yanbing-j) [reply test/test_mkldnn.py]: Sure. Done. I fix it as 2023.

File: torch/_meta_registrations.py (hunk 1 above)
Comment (yanbing-j) [reply torch/_meta_registrations.py]: Yes. An empty tensor works correctly here. Done.

File: aten/src/ATen/native/mkldnn/RNN.cpp (hunk 3 above)
Comment (yanbing-j) [reply aten/src/ATen/native/mkldnn/RNN.cpp]: Remove this replicate.

**Review by jgong5 — APPROVED** [review]

**Comment by yanbing-j** [comment]
This PR depends on Meta internal ideep/oneDNN upgrade. **Do not** merge it before the issue of Meta internal ideep/oneDNN upgrade is fixed.

**Comment by yanbing-j** [comment]
Hi @malfet , could you please help review this PR? Thanks!

File: aten/src/ATen/native/mkldnn/RNN.cpp [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+  auto b = get_mkldnn_tensor(
+      bias, rnn.bias_desc(get_mkldnn_dtype(bias.scalar_type())));
+  auto y = get_mkldnn_tensor(
+      output, rnn.dst_layer_desc(get_mkldnn_dtype(output.scalar_type())));
+  auto hy = get_mkldnn_tensor(
+      hy_, rnn.dst_iter_desc(get_mkldnn_dtype(hy_.scalar_type())));
+  auto cy = get_mkldnn_tensor(
+      cy_, rnn.dst_iter_c_desc(get_mkldnn_dtype(cy_.scalar_type())));
+  auto w1_ = get_mkldnn_tensor(weight_ih, rnn.weights_layer_desc(input_size, get_mkldnn_dtype(weight_ih.scalar_type())));
+  auto w2_ = get_mkldnn_tensor(weight_hh, rnn.weights_iter_desc(get_mkldnn_dtype(weight_hh.scalar_type())));
```
Comment (malfet) [inline aten/src/ATen/native/mkldnn/RNN.cpp:289]: Nit: please add `get_mkldnn_dtype` convenience function that takes `const Tensor&` as an argument.
Perhaps it'll be nicer to add it as static inline right after https://github.com/pytorch/pytorch/blob/e1bcbbf18cae2cf228111b9da784f1052070c724/aten/src/ATen/native/mkldnn/MKLDNNCommon.h#L12
rather than use lambda, but it's up to you
```suggestion
  auto get_mkldnn_dtype = [](const Tensor& t) { return get_mkldnn_dtype(t.scalar_type()); };
  auto x = get_mkldnn_tensor(
      input,
      rnn.src_layer_desc(input_size, get_mkldnn_dtype(input)));
  auto hx = get_mkldnn_tensor(
      hx_, rnn.src_iter_desc(get_mkldnn_dtype(hx_));
  auto cx = get_mkldnn_tensor(
      cx_, rnn.src_iter_c_desc(get_mkldnn_dtype(cx_)));
  auto b = get_mkldnn_tensor(
      bias, rnn.bias_desc(get_mkldnn_dtype(bias)));
  auto y = get_mkldnn_tensor(
      output, rnn.dst_layer_desc(get_mkldnn_dtype(output)));
  auto hy = get_mkldnn_tensor(
      hy_, rnn.dst_iter_desc(get_mkldnn_dtype(hy_)));
  auto cy = get_mkldnn_tensor(
      cy_, rnn.dst_iter_c_desc(get_mkldnn_dtype(cy_)));
  auto w1_ = get_mkldnn_tensor(weight_ih, rnn.weights_layer_desc(input_size, get_mkldnn_dtype(weight_ih)));
  auto w2_ = get_mkldnn_tensor(weight_hh, rnn.weights_iter_desc(get_mkldnn_dtype(weight_hh)));
```

File: aten/src/ATen/native/mkldnn/RNN.cpp [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+      ? _shuffle_bias(weight2, weight3, rnn.mode)
+      : at::zeros({rnn.num_bias_gates * rnn.hidden_size}, weight_ih.options());
+
+  at::Tensor cx_;
+  if (hx_.storage().unsafeGetStorageImpl() ==
+      cx_tmp.storage().unsafeGetStorageImpl()) {
+    cx_ = at::clone(cx_tmp);
+  } else {
+    cx_ = cx_tmp;
+  }
```
Comment (malfet) [inline aten/src/ATen/native/mkldnn/RNN.cpp:359]: Wouldn't ternary be better there? 
```suggestion
  auto cx_  =  hx_.storage().unsafeGetStorageImpl() == cx_tmp.storage().unsafeGetStorageImpl() ? at::clone(cx_tmp) : cx_tmp;
```

File: aten/src/ATen/native/mkldnn/RNN.cpp [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+
+  MatrixRef<Tensor> weights{weight, static_cast<size_t>(weight_stride0)};
+
+  auto num_directions = bidirectional ? 2 : 1;
+  auto layer_input = input;
+  std::vector<at::Tensor> layer_output(num_directions);
+  std::vector<at::Tensor> layer_hy(num_layers * num_directions);
+  std::vector<at::Tensor> layer_cy(num_layers * num_directions);
+  for (int64_t layer = 0; layer < num_layers; layer++) {
+    for (int64_t direction = 0; direction < num_directions; direction++) {
```
Comment (malfet) [inline aten/src/ATen/native/mkldnn/RNN.cpp:497]: Please use `irange` to express that those values are not mutated inside the loop
```suggestion
  for (const auto layer: c10::irange(num_layers)) {
    for (const auto direction: c10::irange(num_directions)) {
```

File: aten/src/ATen/native/mkldnn/RNN.cpp [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+      auto index = layer * num_directions + direction;
```
Comment (malfet) [inline aten/src/ATen/native/mkldnn/RNN.cpp:498]:
```suggestion
      const auto index = layer * num_directions + direction;
```

**Review by malfet — APPROVED** [review]
Please address comments and add more comments explaining what this code is trying to do, though overall looks fine

File: aten/src/ATen/native/mkldnn/RNN.cpp [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+  auto w1_ = get_mkldnn_tensor(weight_ih, rnn.weights_layer_desc(input_size, get_mkldnn_dtype(weight_ih.scalar_type())));
+  auto w2_ = get_mkldnn_tensor(weight_hh, rnn.weights_iter_desc(get_mkldnn_dtype(weight_hh.scalar_type())));
+
+  Tensor workspace = Tensor();
+  auto pd = ideep::lstm_forward_training::prepare(
+      x, hx, cx, w1_, w2_, b, y, hy, cy, reverse);
+  workspace = at::empty(pd.workspace_desc().get_size() / sizeof(uint8_t), input.options().dtype(at::kByte));
+  ideep::tensor mkldnn_workspace;
+  mkldnn_workspace.init(
+      pd.workspace_desc(), workspace.template data_ptr<uint8_t>());
```
Comment (malfet) [inline aten/src/ATen/native/mkldnn/RNN.cpp:297]: Why `template` keyword is needed here?
```suggestion
      pd.workspace_desc(), workspace.data_ptr<uint8_t>());
```

**Comment by yanbing-j** [comment]
Hi @malfet , I have addressed all the comments. It's much better to use the suggested changes. I will try to merge this PR then.

**Comment by xuzhao9** [comment]
We observe 40~60% speedup in tts_angular model on CPU in TorchBench: https://github.com/pytorch/benchmark/issues/1376. Congrats!
